diff --git a/test/test-list.h b/test/test-list.h
index b89930d..59b799f 100644
--- a/test/test-list.h
+++ b/test/test-list.h
@@ -46,7 +46,6 @@ TEST_DECLARE   (condvar_6)
 TEST_DECLARE   (semaphore_1)
 TEST_DECLARE   (semaphore_2)
 TEST_DECLARE   (semaphore_3)
-TEST_DECLARE   (tty)
 #ifdef _WIN32
 TEST_DECLARE   (tty_raw)
 TEST_DECLARE   (tty_empty_write)
@@ -468,7 +467,6 @@ TASK_LIST_START
 #endif
   TEST_ENTRY  (pipe_set_non_blocking)
   TEST_ENTRY  (pipe_set_chmod)
-  TEST_ENTRY  (tty)
 #ifdef _WIN32
   TEST_ENTRY  (tty_raw)
   TEST_ENTRY  (tty_empty_write)
