Index: matplotlib-3.3.2/src/_macosx.m
===================================================================
--- matplotlib-3.3.2.orig/src/_macosx.m
+++ matplotlib-3.3.2/src/_macosx.m
@@ -1207,10 +1207,12 @@ NavigationToolbar2_init(NavigationToolba
     rect.size.height = 0;
     rect.origin.x += height;
     NSTextView* messagebox = [[NSTextView alloc] initWithFrame: rect];
+#if 0
     if (@available(macOS 10.11, *)) {
         messagebox.textContainer.maximumNumberOfLines = 2;
         messagebox.textContainer.lineBreakMode = NSLineBreakByTruncatingTail;
     }
+#endif
     [messagebox setFont: font];
     [messagebox setDrawsBackground: NO];
     [messagebox setSelectable: NO];
