--- work/src/gmpy_xmpz.c.orig	2018-08-02 16:58:21.417395100 -0500
+++ work/src/gmpy_xmpz.c	2018-08-02 16:58:32.730617500 -0500
@@ -341,12 +341,15 @@
 
 #if PY_VERSION_HEX > 0x030200A4
         if (PySlice_GetIndicesEx(item,
+                        seq_len,
+                        &start, &stop, &step, &slicelength) < 0) {
+            return -1;
 #else
         if (PySlice_GetIndicesEx((PySliceObject*)item,
-#endif
                         seq_len,
                         &start, &stop, &step, &slicelength) < 0) {
             return -1;
+#endif
         }
 
         if ((step < 0 && start < stop) ||
