--- pthreads.2/pthread.h
+++ pthreads.2/pthread.h
@@ -105,6 +105,7 @@
 #define PTW32_LEVEL_MAX 3
 
 #if ( defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 200112 )  || !defined(PTW32_LEVEL)
+#undef PTW32_LEVEL
 #define PTW32_LEVEL PTW32_LEVEL_MAX
 /* Include everything */
 #endif
--- pthreads.2/sched.h
+++ pthreads.2/sched.h
@@ -61,6 +61,7 @@
 #define PTW32_SCHED_LEVEL_MAX 3
 
 #if ( defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 200112 )  || !defined(PTW32_SCHED_LEVEL)
+#undef PTW32_SCHED_LEVEL
 #define PTW32_SCHED_LEVEL PTW32_SCHED_LEVEL_MAX
 /* Include everything */
 #endif

