diff --git a/llvmlite/binding/ffi.py b/llvmlite/binding/ffi.py
index 3f53b2a..cb34899 100644
--- a/llvmlite/binding/ffi.py
+++ b/llvmlite/binding/ffi.py
@@ -119,6 +119,9 @@ if os.name == 'nt':
     # Append DLL directory to PATH, to allow loading of bundled CRT libraries
     # (Windows uses PATH for DLL loading, see http://msdn.microsoft.com/en-us/library/7d83bc18.aspx).  # noqa E501
     os.environ['PATH'] += ';' + _lib_dir
+    import sys
+    if sys.version_info >= (3, 8):
+        os.add_dll_directory(os.path.join(os.path.normpath(sys.prefix), 'Library', 'bin'))
 
 
 _lib_name = get_library_name()
