From 8afe829a8bb877edcd9da9580ef8529761f9b1cb Mon Sep 17 00:00:00 2001
From: Nehal J Wani <nehaljw.kkd1@gmail.com>
Date: Sat, 18 Nov 2017 10:14:47 -0600
Subject: [PATCH] Fix name of static sodium lib

---
 src/bindings/build.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/bindings/build.py b/src/bindings/build.py
index a62ba42..1c58fb9 100644
--- a/src/bindings/build.py
+++ b/src/bindings/build.py
@@ -47,7 +47,7 @@ if os.getenv("PYNACL_SODIUM_STATIC") is not None:
 source.append("#include <sodium.h>")
 
 if sys.platform == "win32":
-    libraries = ["libsodium"]
+    libraries = ["libsodium_static"]
 else:
     libraries = ["sodium"]
 
-- 
2.7.4

